blob/azureblob/azureblob.go

```diff
@@ -816,14 +821,54 @@ func (b *bucket) ListPaged(ctx context.Context, opts *driver.ListOptions) (*driv
 	return page, nil
 }
 
+func (b *bucket) getDelegationCredentials(ctx context.Context) (azblob.StorageAccountCredential, error) {
+	validPeriod := 48 * time.Hour
+	currentTime := time.Now().UTC()
+	expires := currentTime.Add(validPeriod)
+	b.credentialExpiration = &expires
+	keyInfo := azblob.NewKeyInfo(currentTime, expires)
+	delegationCredentials, err := b.serviceURL.GetUserDelegationCredential(ctx, keyInfo, nil, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("error while retrieving user delegation credential: %s", err)
+	}
+
+	return delegationCredentials, nil
+}
+
+func (b *bucket) refreshDelegationCredentials(ctx context.Context) (azblob.StorageAccountCredential, error) {
+	var err error
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+	creds := b.delegationCredentials
+
+	if b.credentialExpiration != nil && time.Now().UTC().After(*b.credentialExpiration) {
+		creds, err = b.getDelegationCredentials(ctx)
+		b.delegationCredentials = creds
+	}
+	return creds, err
+}
+
 // SignedURL implements driver.SignedURL.
 func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
-	if b.opts.Credential == nil {
+	var credential azblob.StorageAccountCredential
+	isMSIEnvironment := adal.MSIAvailable(ctx, adal.CreateSender())
```

Comment:
Nit: might as well do this inline on line 858 to avoid doing it if b.opts.Credential is provided.